### PR TITLE
Add support for parsing breakpad sym files containing INLINE_ORIGIN and INLINE records

### LINKED ISSUE #432: Add support for `INLINE` and `INLINE_ORIGIN` entries when parsing .sym files
Google added support for inlined functions in Breakpad via the `INLINE` and `INLINE_ORIGIN` entries (see [this commit](https://chromium.googlesource.com/breakpad/breakpad/+/4f5b814790301e7b8252535e0732bee11f4bb801) and [this commit](https://chromium.googlesource.com/breakpad/breakpad/+/f0803507950279bd1e48d05f5fb4b4377c2917f7)). Unfortunately they didn't update the documentation. This bug is about supporting them in symbolic-debuginfo.

The records look like this:

`INLINE_ORIGIN <origin_id> <name>`

* `origin_id` identifies the inlined function, Breakpad uses a counter starting at 0 for these IDs and it must be unique
* `name` is the name of the inlined function

`INLINE <inline_nest_level> <call_site_line> <call_site_file_id> <origin_id> <address> <size>`

* `inline_nest_level` is the nesting level, starting at 0 for the outermost function
* `call_site_line` is the line at the call site
* `call_site_file_id` points to the `FILE` entry where this call has been inlined
* `origin_id` points to the corresponding `INLINE_ORIGIN` entry
* `address` is the starting address in the module
* `size` is the function size in the module

`INLINE` entries appear after each `FUNC` entry, describing the functions that have been inlined in there, e.g.

```
FUNC 1000 20 0 foo
INLINE 0 ...
INLINE 1 ...
(more INLINE entries) ...
1000 ... // regular source line data
(more source line entries) ...
FUNC 1020 40 0 bar
INLINE 0 ...
INLINE 1 ...
INLINE 2 ...
1020 ... // regular source line data
etc...
```

One thing worth keeping in mind is that prior to this change .sym files used to have a specific structures WRT the order in which the various entries appeared. This does not appear to be true anymore, specifically [this](https://chromium.googlesource.com/breakpad/breakpad/+/e87bb1b3b480d1e44e6a742b8cc0a1698b98494e) points to the fact that `INLINE_ORIGIN` entries can be mixed up with `FUNC` ones.

### PR DESCRIPTION
**Pull request by mstange**
Fixes #432.

Please review commits individually.

### CODE REVIEW

**Review by Swatinem — APPROVED** [review]
This looks very nice!

There is a bunch of asserts and unwraps, I wonder how resilient this is to malformed breakpad files? Is it worth the effort to run this through a fuzzer? Or proptest that programmatically generates inlinee and line records?

File: symbolic-debuginfo/src/breakpad.rs [hunk 1]
```diff
@@ -614,6 +644,69 @@ impl<'d> Iterator for BreakpadLineRecords<'d> {
...
+        file_map.get(&self.call_site_file_id).cloned()
+    }
+
+    /// Resolves the function name for this record in the inline origin map.
+    pub fn function_name<'d>(
+        &self,
+        inline_origin_map: &BreakpadInlineOriginMap<'d>,
+    ) -> Option<&'d str> {
+        inline_origin_map.get(&self.origin_id).cloned()
+    }
```
Comment (loewenheim) [inline symbolic-debuginfo/src/breakpad.rs:690]: Honestly, I would inline these functions. They are each only used once.

**Review by loewenheim — APPROVED** [review]
Looks great. What's your reasoning for adding stuff to `shared.rs`? AFAICT it's only used for Breakpad processing.

**Comment by loewenheim** [comment]
> There is a bunch of asserts and unwraps, I wonder how resilient this is to malformed breakpad files? Is it worth the effort to run this through a fuzzer? Or proptest that programmatically generates inlinee and line records?

I'm pretty sure all of these are actually infallible, irrespective of whether the Breakpad file is well-formed. They only unwrap if it has been checked immediately before that an element exists. Some of this stuff could arguably be rewritten in terms of matching, but I'm not sure it's worth the hassle.

**Comment by loewenheim** [comment]
Also I opened https://github.com/getsentry/symbolic/pull/609 to use `saturating_add` in `Function::end_address`, so you can use that method if you like. We can also do it as follow-up.

**Comment by Swatinem** [comment]
Fair enough, I looked again at the FunctionBuilder, and all the unwrap/assert look to be unreachable. I wonder if the compiler is smart enough to optimize all that away as well.

I also agree that the `FunctionBuilder` should maybe move into its own file, so we can get rid of a bunch of `cfg` switches.

**Comment by mstange** [comment]
Thanks for the fast reviews! I'll open a new PR for just the first commit, as Arpad suggested.

Creating a new file for `FunctionBuilder` sounds good, I'll do that. I didn't put it into breakpad.rs because #607 adds another use of it in the DWARF parser.

File: symbolic-debuginfo/src/breakpad.rs (hunk 1 above)
Comment (mstange) [reply symbolic-debuginfo/src/breakpad.rs:690]: Sounds good. I created `BreakpadInlineRecord` by copying `BreakpadLineRecord` which already has a similar method for the `filename`, but I agree there's not much value in having those methods.

**Comment by mstange** [comment]
I've addressed the review comments.

**Comment by mstange** [comment]
> I wonder how resilient this is to malformed breakpad files? Is it worth the effort to run this through a fuzzer?

It might be worth it, just for added confidence. But I'm not too concerned about the unwraps and asserts, as you already said I think they're all unreachable.

There's one unchecked addition which could overflow on malicious input on 32 bit platforms (and panic debug builds), and that's the one in `flush_depth`:

```rust
    pub fn flush_depth(&mut self, depth: u32) {
        while self.stack.len() > depth as usize + 1 {
```

I've chosen to not care about it, but I can fix it if you'd like me to.

**Review by Swatinem — APPROVED** [review]

**Pull request merged**